src/main/java/mate/academy/dao/BookDaoImpl.java

```diff
@@ -0,0 +1,117 @@
+            }
+            ResultSet generatedKeys = statement.getGeneratedKeys();
+            if (generatedKeys.next()) {
+                Long id = generatedKeys.getObject(1, Long.class);
+                book.setId(id);
+            }
+            return book;
+        } catch (SQLException e) {
+            throw new DataProcessingException(ERROR + " creation - " + book, e);
+        }
+    }
+
+    @Override
+    public List<Book> findAll() {
+        List<Book> response = new ArrayList<>();
+        try (Connection connection = ConnectionUtil.getConnection();
+                 PreparedStatement statement = connection.prepareStatement(FIND_ALL_QUERY)) {
+            ResultSet resultSet = statement.executeQuery();
+            while (resultSet.next()) {
+                response.add(getBook(resultSet));
+            }
+            return response;
+        } catch (SQLException e) {
+            throw new DataProcessingException(ERROR + " finding all books", e);
+        }
+    }
+
+    @Override
+    public Optional<Book> findById(Long id) {
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(FIND_BY_ID_QUERY)) {
+            statement.setLong(1, id);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                return Optional.of(getBook(resultSet));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException(ERROR + " finding by id = " + id, e);
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public Book update(Book book) {
+        try (Connection connection = ConnectionUtil.getConnection();
+                 PreparedStatement statement = connection.prepareStatement(UPDATE_QUERY);) {
+            statement.setString(1, book.getTitle());
+            statement.setBigDecimal(2, book.getPrice());
+            statement.setLong(3, book.getId());
+            int rowsUpdated = statement.executeUpdate();
+            if (rowsUpdated < 1) {
+                throw new DataProcessingException("Failed to update book, no rows were affected");
+            }
+            return book;
+        } catch (SQLException e) {
+            throw new DataProcessingException(ERROR + " updating - " + book, e);
+        }
+    }
+
+    @Override
+    public boolean delete(Book book) {
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(DELETE_QUERY)) {
+            statement.setLong(1, book.getId());
+            int rowsUpdated = statement.executeUpdate();
+            return rowsUpdated > 0;
+        } catch (SQLException e) {
+            throw new RuntimeException(ERROR + " deleting - " + book, e);
+        }
+    }
+
+    private Book getBook(ResultSet resultSet) throws SQLException {
+        Long id = resultSet.getLong(ID);
```

Comment:
You should be more responsible with the common mistakes list, usually around half or more of the ones you make - are present there, and it's the second time we are pointing this one out, you have to check all the similar places, and not just the one where we left the comment
![image](https://github.com/mate-academy/jv-jdbc-intro/assets/80981426/add9c453-62bd-44af-a87b-7e64a1f98d3c)
Similar think was about the code duplication